File: app/src/main/java/com/yhjoo/dochef/activities/HomeActivity.java
package com.yhjoo.dochef.activities;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.text.InputType;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.afollestad.materialdialogs.MaterialDialog;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.theartofdev.edmodo.cropper.CropImage;
import com.theartofdev.edmodo.cropper.CropImageView;
import com.yhjoo.dochef.App;
import com.yhjoo.dochef.R;
import com.yhjoo.dochef.adapter.PostListAdapter;
import com.yhjoo.dochef.adapter.RecipeHorizontalHomeAdapter;
import com.yhjoo.dochef.databinding.AHomeBinding;
import com.yhjoo.dochef.interfaces.RxRetrofitServices;
import com.yhjoo.dochef.model.Post;
import com.yhjoo.dochef.model.Recipe;
import com.yhjoo.dochef.model.UserDetail;
import com.yhjoo.dochef.utils.DataGenerator;
import com.yhjoo.dochef.utils.GlideApp;
import com.yhjoo.dochef.utils.ImageLoadUtil;
import com.yhjoo.dochef.utils.RxRetrofitBuilder;
import com.yhjoo.dochef.utils.Utils;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.functions.Function;
import retrofit2.Response;

public class HomeActivity extends BaseActivity {
    private final int CODE_PERMISSION = 22;
    private final int IMG_WIDTH = 360;
    private final int IMG_HEIGHT = 360;

    private enum MODE {OWNER, OTHERS}

    private enum OPERATION {VIEW, REVISE}

    AHomeBinding binding;
    StorageReference storageReference;
    RxRetrofitServices.AccountService accountService;
    RxRetrofitServices.UserService userService;
    RxRetrofitServices.RecipeService recipeService;
    RxRetrofitServices.PostService postService;
    RecipeHorizontalHomeAdapter recipeHorizontalHomeAdapter;
    PostListAdapter postListAdapter;

    MenuItem reviseMenu;
    MenuItem okMenu;

    ArrayList<Recipe> recipeList;
    ArrayList<Post> postList;
    UserDetail userDetailInfo;

    MODE currentMode;
    OPERATION currentOperation = OPERATION.VIEW;
    Uri mImageUri;
    String image_url;
    String currentUserID;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = AHomeBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.homeToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        storageReference = FirebaseStorage.getInstance().getReference();

        accountService = RxRetrofitBuilder.create(this, RxRetrofitServices.AccountService.class);
        userService = RxRetrofitBuilder.create(this, RxRetrofitServices.UserService.class);
        recipeService = RxRetrofitBuilder.create(this, RxRetrofitServices.RecipeService.class);
        postService = RxRetrofitBuilder.create(this, RxRetrofitServices.PostService.class);

        String userID = Utils.getUserBrief(this).getUserID();
        if (getIntent().getStringExtra("userID") == null
                || getIntent().getStringExtra("userID").equals(userID)) {
            currentMode = MODE.OWNER;
            currentUserID = userID;
        } else {
            currentMode = MODE.OTHERS;
            currentUserID = getIntent().getStringExtra("userID");
        }

        recipeHorizontalHomeAdapter = new RecipeHorizontalHomeAdapter(currentUserID);
        recipeHorizontalHomeAdapter.setOnItemClickListener((adapter, view, position) -> {
            Intent intent = new Intent(HomeActivity.this, RecipeDetailActivity.class)
                    .putExtra("recipeID", recipeList.get(position).getRecipeID());
            startActivity(intent);
        });
        binding.homeRecipeRecycler.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));
        binding.homeRecipeRecycler.setAdapter(recipeHorizontalHomeAdapter);

        postListAdapter = new PostListAdapter();
        postListAdapter.setOnItemClickListener((adapter, view, position) -> {
            Intent intent = new Intent(HomeActivity.this, PostDetailActivity.class)
                    .putExtra("postID", postList.get(position).getPostID());
            startActivity(intent);
        });
        binding.homePostRecycler.setLayoutManager(new LinearLayoutManager(this) {
            @Override
            public boolean canScrollHorizontally() {
                return false;
            }

            @Override
            public boolean canScrollVertically() {
                return false;
            }
        });
        binding.homePostRecycler.setAdapter(postListAdapter);

        if (App.isServerAlive()) {
            loadList();
        } else {
            userDetailInfo = DataGenerator.make(getResources(), getResources().getInteger(R.integer.DATA_TYPE_USER_DETAIL));
            recipeList = DataGenerator.make(getResources(), getResources().getInteger(R.integer.DATE_TYPE_RECIPE));
            postList = DataGenerator.make(getResources(), getResources().getInteger(R.integer.DATA_TYPE_POST));

            setUserInfo();
            recipeHorizontalHomeAdapter.setNewData(recipeList);
            postListAdapter.setNewData(postList);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (currentMode == MODE.OWNER) {
            getMenuInflater().inflate(R.menu.menu_home_owner, menu);

            reviseMenu = menu.findItem(R.id.menu_home_owner_revise);
            okMenu = menu.findItem(R.id.menu_home_owner_revise_ok);
        }

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onBackPressed() {
        if (currentOperation == OPERATION.REVISE) {
            createConfirmDialog(this,
                    null, "변경이 취소됩니다.",
                    (dialog1, which) -> {
                        currentOperation = OPERATION.VIEW;
                        reviseMenu.setVisible(true);
                        okMenu.setVisible(false);
                        binding.homeRevisegroup.setVisibility(View.GONE);

                        ImageLoadUtil.loadUserImage(this, userDetailInfo.getUserImg(), binding.homeUserimg);
                        binding.homeNickname.setText(userDetailInfo.getNickname());
                        binding.homeProfiletext.setText(userDetailInfo.getProfileText());
                    }
            )
                    .show();
        } else
            super.onBackPressed();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.menu_home_owner_revise) {
            currentOperation = OPERATION.REVISE;
            reviseMenu.setVisible(false);
            okMenu.setVisible(true);
            binding.homeRevisegroup.setVisibility(View.VISIBLE);
        } else if (item.getItemId() == R.id.menu_home_owner_revise_ok) {
            updateProfile();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            CropImage.ActivityResult result = CropImage.getActivityResult(data);
            if (resultCode == RESULT_OK) {
                mImageUri = result.getUri();

                Utils.log(result.getUri());
                GlideApp.with(this)
                        .load(mImageUri)
                        .circleCrop()
                        .diskCacheStrategy(DiskCacheStrategy.NONE)
                        .skipMemoryCache(true)
                        .into(binding.homeUserimg);

            } else if (resultCode == CropImage.CROP_IMAGE_ACTIVITY_RESULT_ERROR_CODE) {
                Exception error = result.getError();
                error.printStackTrace();
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == CODE_PERMISSION) {
            for (int result : grantResults)
                if (result == PackageManager.PERMISSION_DENIED) {
                    App.getAppInstance().showToast("권한 거부");
                    return;
                }

            CropImage.activity(mImageUri)
                    .setGuidelines(CropImageView.Guidelines.ON)
                    .setAspectRatio(1, 1)
                    .setCropShape(CropImageView.CropShape.OVAL)
                    .setRequestedSize(IMG_WIDTH, IMG_HEIGHT)
                    .setMaxCropResultSize(IMG_WIDTH, IMG_HEIGHT)
                    .setOutputUri(mImageUri)
                    .start(this);
        }
    }

    void loadList() {
        compositeDisposable.add(
                userService.getUserDetail(currentUserID)
                        .flatMap((Function<Response<UserDetail>, Single<Response<ArrayList<Recipe>>>>)
                                response -> {
                                    userDetailInfo = response.body();
                                    return recipeService.getRecipeByUserID(currentUserID, "latest");
                                })
                        .flatMap((Function<Response<ArrayList<Recipe>>, Single<Response<ArrayList<Post>>>>)
                                response -> {
                                    List<Recipe> res = response.body().subList(0, Math.min(response.body().size(), 10));
                                    recipeList = new ArrayList<>(res);

                                    return postService.getPostListByUserID(currentUserID)
                                            .observeOn(AndroidSchedulers.mainThread());
                                })
                        .subscribe(response -> {
                            postList = response.body();

                            setUserInfo();

                            recipeHorizontalHomeAdapter.setNewData(recipeList);
                            recipeHorizontalHomeAdapter.setEmptyView(
                                    R.layout.rv_empty_recipe, (ViewGroup) binding.homeRecipeRecycler.getParent());

                            postListAdapter.setNewData(postList);
                            postListAdapter.setEmptyView(
                                    R.layout.rv_empty_post, (ViewGroup) binding.homePostRecycler.getParent());
                        }, RxRetrofitBuilder.defaultConsumer())
        );
    }

    void setUserInfo() {
        ImageLoadUtil.loadUserImage(this, userDetailInfo.getUserImg(), binding.homeUserimg);

        binding.homeToolbar.setTitle(userDetailInfo.getNickname());
        binding.homeNickname.setText(userDetailInfo.getNickname());
        binding.homeProfiletext.setText(userDetailInfo.getProfileText());
        binding.homeRecipecount.setText(Integer.toString(userDetailInfo.getRecipeCount()));
        binding.homeFollowercount.setText(Integer.toString(userDetailInfo.getFollowerCount()));
        binding.homeFollowingcount.setText(Integer.toString(userDetailInfo.getFollowingCount()));

        binding.homeFollowBtn.setVisibility(currentMode == MODE.OTHERS ? View.VISIBLE : View.GONE);

        binding.homeUserimgRevise.setOnClickListener(this::reviseProfileImage);
        binding.homeNicknameRevise.setOnClickListener(this::clickReviseNickname);
        binding.homeProfiletextRevise.setOnClickListener(this::clickReviseContents);
        binding.homeRecipewrapper.setOnClickListener((v -> {
            Intent intent = new Intent(HomeActivity.this, RecipeMyListActivity.class)
                    .putExtra("userID", userDetailInfo.getUserID());
            startActivity(intent);
        }));
        binding.homeFollowerwrapper.setOnClickListener((v -> {
            Intent intent = new Intent(HomeActivity.this, FollowListActivity.class)
                    .putExtra("MODE", FollowListActivity.MODE.FOLLOWER)
                    .putExtra("userID", userDetailInfo.getUserID());
            startActivity(intent);
        }));
        binding.homeFollowingwrapper.setOnClickListener((v -> {
            Intent intent = new Intent(HomeActivity.this, FollowListActivity.class)
                    .putExtra("MODE", FollowListActivity.MODE.FOLLOWING)
                    .putExtra("userID", userDetailInfo.getUserID());
            startActivity(intent);
        }));
    }

    void reviseProfileImage(View v) {
        final String[] permissions = {
                Manifest.permission.READ_EXTERNAL_STORAGE
        };

        if (Utils.checkPermission(this, permissions)) {
            CropImage.activity()
                    .setGuidelines(CropImageView.Guidelines.ON)
                    .setAspectRatio(1, 1)
                    .setOutputUri(mImageUri)
                    .setCropShape(CropImageView.CropShape.OVAL)
                    .setRequestedSize(IMG_WIDTH, IMG_HEIGHT)
                    .setMaxCropResultSize(IMG_WIDTH, IMG_HEIGHT)
                    .start(this);
        } else
            ActivityCompat.requestPermissions(this, permissions, CODE_PERMISSION);
    }

    void clickReviseNickname(View v) {
        MaterialDialog materialDialog = new MaterialDialog.Builder(this)
                .autoDismiss(false)
                .title("닉네임 변경")
                .inputType(InputType.TYPE_CLASS_TEXT)
                .inputRange(4, 12)
                .input("닉네임", binding.homeNickname.getText().toString(), (dialog, input) -> {
                })
                .positiveText("확인")
                .positiveColorRes(R.color.grey_text)
                .onPositive((dialog, which) -> {
                    if (dialog.getInputEditText().getText() == null) {
                        App.getAppInstance().showToast("닉네임을 입력 해 주세요.");
                    } else if (dialog.getInputEditText().getText().toString().length() > 12) {
                        App.getAppInstance().showToast("12자 이하 입력 해 주세요.");
                    } else {
                        compositeDisposable.add(
                                accountService.checkNickname(dialog.getInputEditText().getText().toString())
                                        .observeOn(AndroidSchedulers.mainThread())
                                        .subscribe(response -> {
                                            String msg = response.body().get("msg").getAsString();
                                            Utils.log(msg);
                                            if (msg.equals("ok")) {
                                                binding.homeNickname.setText(dialog.getInputEditText().getText().toString());
                                                dialog.dismiss();
                                            } else
                                                App.getAppInstance().showToast("이미 존재합니다.");
                                        }, RxRetrofitBuilder.defaultConsumer())
                        );
                    }
                })
                .negativeText("취소")
                .negativeColorRes(R.color.grey_text)
                .onNegative((dialog, which) -> dialog.dismiss())
                .build();

        materialDialog.show();
    }

    void clickReviseContents(View v) {
        MaterialDialog materialDialog = new MaterialDialog.Builder(this)
                .autoDismiss(false)
                .title("프로필 변경")
                .inputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_FLAG_MULTI_LINE)
                .inputRange(0, 60)
                .input("프로필", binding.homeProfiletext.getText().toString(), (dialog, input) -> {
                })
                .positiveText("확인")
                .positiveColorRes(R.color.grey_text)
                .onPositive((dialog, which) -> {
                    if (dialog.getInputEditText().getText() == null) {
                        App.getAppInstance().showToast("프로필을 입력 해 주세요.");
                    } else if (dialog.getInputEditText().getText().toString().length() > 60) {
                        App.getAppInstance().showToast("60자 이하 입력 해 주세요.");
                    } else if (dialog.getInputEditText().getLineCount() > 4) {
                        App.getAppInstance().showToast("4줄 이하 입력 해 주세요.");
                    } else {
                        binding.homeProfiletext.setText(dialog.getInputEditText().getText().toString());
                        dialog.dismiss();
                    }
                })
                .negativeText("취소")
                .negativeColorRes(R.color.grey_text)
                .onNegative((dialog, which) -> dialog.dismiss())
                .build();

        materialDialog.show();
    }

    void updateProfile() {
        progressON(this);

        image_url = "";
        if (mImageUri != null) {
            image_url = String.format(getString(R.string.format_upload_file),
                    currentUserID, Long.toString(System.currentTimeMillis()));
            StorageReference ref = storageReference.child(getString(R.string.storage_path_profile) + image_url);
            ref.putFile(mImageUri)
                    .addOnSuccessListener(taskSnapshot -> updateToServer());
        } else{
            image_url = userDetailInfo.getUserImg();
            updateToServer();
        }

    }

    void updateToServer() {
        compositeDisposable.add(
                accountService.updateUser(userDetailInfo.getUserID(), image_url,
                        binding.homeNickname.getText().toString(),
                        binding.homeProfiletext.getText().toString())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(response -> {
                            App.getAppInstance().showToast("업데이트 되었습니다.");

                            currentOperation = OPERATION.VIEW;
                            reviseMenu.setVisible(true);
                            okMenu.setVisible(false);
                            binding.homeRevisegroup.setVisibility(View.VISIBLE);

                            mImageUri = null;
                            progressOFF();
                        }, RxRetrofitBuilder.defaultConsumer())
        );
    }
}

File: app/src/main/java/com/yhjoo/dochef/ui/activities/HomeActivity.kt
package com.yhjoo.dochef.ui.activities

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.text.InputType
import android.view.*
import androidx.core.app.ActivityCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.getInputField
import com.afollestad.materialdialogs.input.input
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.canhub.cropper.CropImage
import com.canhub.cropper.CropImageView
import com.chad.library.adapter.base.BaseQuickAdapter
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.yhjoo.dochef.App
import com.yhjoo.dochef.R
import com.yhjoo.dochef.db.DataGenerator
import com.yhjoo.dochef.model.Post
import com.yhjoo.dochef.model.Recipe
import com.yhjoo.dochef.model.UserDetail
import com.yhjoo.dochef.databinding.AHomeBinding
import com.yhjoo.dochef.ui.adapter.PostListAdapter
import com.yhjoo.dochef.ui.adapter.RecipeHorizontalHomeAdapter
import com.yhjoo.dochef.utilities.*
import com.yhjoo.dochef.utilities.RetrofitServices.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.*

class HomeActivity : BaseActivity() {
    companion object {
        const val CODE_PERMISSION = 22
        const val IMG_WIDTH = 360
        const val IMG_HEIGHT = 360
    }

    object UIMODE {
        const val OWNER = 0
        const val OTHERS = 1
    }

    object OPERATION {
        const val VIEW = 0
        const val REVISE = 1
    }

    val binding: AHomeBinding by lazy { AHomeBinding.inflate(layoutInflater) }

    private lateinit var storageReference: StorageReference
    private lateinit var accountService: AccountService
    private lateinit var userService: UserService
    private lateinit var recipeService: RecipeService
    private lateinit var postService: PostService

    private lateinit var recipeHorizontalHomeAdapter: RecipeHorizontalHomeAdapter
    private lateinit var postListAdapter: PostListAdapter

    private lateinit var reviseMenu: MenuItem
    private lateinit var okMenu: MenuItem

    private lateinit var recipeList: ArrayList<Recipe>
    private lateinit var postList: ArrayList<Post>
    private lateinit var userDetailInfo: UserDetail

    private var currentMode: Int? = null
    private var currentOperation: Int? = null

    private var imageUri: Uri? = null
    private var imageUrl: String? = null
    private var currentUserID: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        setSupportActionBar(binding.homeToolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        storageReference = FirebaseStorage.getInstance().reference

        accountService = RetrofitBuilder.create(this, AccountService::class.java)
        userService = RetrofitBuilder.create(this, UserService::class.java)
        recipeService = RetrofitBuilder.create(this, RecipeService::class.java)
        postService = RetrofitBuilder.create(this, PostService::class.java)

        val userID = Utils.getUserBrief(this).userID
        if (intent.getStringExtra("userID") == null || intent.getStringExtra("userID") == userID) {
            currentMode = UIMODE.OWNER
            currentUserID = userID
        } else {
            currentMode = UIMODE.OTHERS
            currentUserID = intent.getStringExtra("userID")
        }

        recipeHorizontalHomeAdapter = RecipeHorizontalHomeAdapter(currentUserID)
        recipeHorizontalHomeAdapter.setOnItemClickListener { _: BaseQuickAdapter<*, *>?, _: View?, position: Int ->
            val intent = Intent(this@HomeActivity, RecipeDetailActivity::class.java)
                .putExtra("recipeID", recipeList[position].recipeID)
            startActivity(intent)
        }

        binding.apply {
            homeRecipeRecycler.layoutManager =
                LinearLayoutManager(this@HomeActivity, LinearLayoutManager.HORIZONTAL, false)
            homeRecipeRecycler.adapter = recipeHorizontalHomeAdapter

            postListAdapter = PostListAdapter()
            postListAdapter.setOnItemClickListener { _: BaseQuickAdapter<*, *>?, _: View?, position: Int ->
                val intent = Intent(this@HomeActivity, PostDetailActivity::class.java)
                    .putExtra("postID", postList[position].postID)
                startActivity(intent)
            }

            binding.homePostRecycler.layoutManager =
                object : LinearLayoutManager(this@HomeActivity) {
                    override fun canScrollHorizontally(): Boolean {
                        return false
                    }

                    override fun canScrollVertically(): Boolean {
                        return false
                    }
                }
            binding.homePostRecycler.adapter = postListAdapter
        }

        if (App.isServerAlive) {
            loadList()
        } else {
            userDetailInfo =
                DataGenerator.make(resources, resources.getInteger(R.integer.DATA_TYPE_USER_DETAIL))
            recipeList =
                DataGenerator.make(resources, resources.getInteger(R.integer.DATE_TYPE_RECIPE))
            postList = DataGenerator.make(resources, resources.getInteger(R.integer.DATA_TYPE_POST))

            setUserInfo()

            recipeHorizontalHomeAdapter.setNewData(recipeList)
            postListAdapter.setNewData(postList)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (currentMode == UIMODE.OWNER) {
            menuInflater.inflate(R.menu.menu_home_owner, menu)
            reviseMenu = menu.findItem(R.id.menu_home_owner_revise)
            okMenu = menu.findItem(R.id.menu_home_owner_revise_ok)
        }

        return super.onCreateOptionsMenu(menu)
    }

    override fun onBackPressed() {
        if (currentOperation == OPERATION.REVISE) {
            MaterialDialog(this).show {
                positiveButton(text = "확인") {
                    currentOperation = OPERATION.VIEW
                    reviseMenu.isVisible = true
                    okMenu.isVisible = false
                    binding.apply {
                        homeRevisegroup.visibility = View.GONE
                        GlideImageLoadDelegator.loadUserImage(
                            this@HomeActivity,
                            userDetailInfo.userImg,
                            homeUserimg
                        )
                        homeNickname.text = userDetailInfo.nickname
                        homeProfiletext.text = userDetailInfo.profileText
                    }
                }
                negativeButton(text = "취소")
            }
        } else super.onBackPressed()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_home_owner_revise -> {
                currentOperation = OPERATION.REVISE
                reviseMenu.isVisible = false
                okMenu.isVisible = true
                binding.homeRevisegroup.visibility = View.VISIBLE
                true
            }
            R.id.menu_home_owner_revise_ok -> {
                updateProfile()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            val result = CropImage.getActivityResult(data)
            if (resultCode == RESULT_OK) {
                imageUri = result!!.originalUri
                Utils.log(imageUri.toString())
                GlideApp.with(this)
                    .load(imageUri)
                    .circleCrop()
                    .diskCacheStrategy(DiskCacheStrategy.NONE)
                    .skipMemoryCache(true)
                    .into(binding.homeUserimg)
            } else if (resultCode == CropImage.CROP_IMAGE_ACTIVITY_RESULT_ERROR_CODE) {
                val error = result!!.error
                error!!.printStackTrace()
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == CODE_PERMISSION) {
            for (result in grantResults) if (result == PackageManager.PERMISSION_DENIED) {
                App.showToast("권한 거부")
                return
            }
            CropImage.activity(imageUri)
                .setGuidelines(CropImageView.Guidelines.ON)
                .setAspectRatio(1, 1)
                .setCropShape(CropImageView.CropShape.OVAL)
                .setRequestedSize(IMG_WIDTH, IMG_HEIGHT)
                .setMaxCropResultSize(IMG_WIDTH, IMG_HEIGHT)
                .setOutputUri(imageUri)
                .start(this)
        }
    }

    private fun loadList() {
        CoroutineScope(Dispatchers.Main).launch {
            runCatching {
                val res1 = userService.getUserDetail(currentUserID!!)
                userDetailInfo = res1.body()!!

                val res2 = recipeService.getRecipeByUserID(currentUserID!!, "latest")
                val res2Data: List<Recipe?> = res2.body()!!
                    .subList(0, res2.body()!!.size.coerceAtMost(10))
                recipeList = ArrayList(res2Data)

                val res3 = postService.getPostListByUserID(currentUserID!!)
                postList = res3.body()!!
                setUserInfo()

                recipeHorizontalHomeAdapter.setNewData(recipeList)
                recipeHorizontalHomeAdapter.setEmptyView(
                    R.layout.rv_empty_recipe, binding.homeRecipeRecycler.parent as ViewGroup
                )
                postListAdapter.setNewData(postList)
                postListAdapter.setEmptyView(
                    R.layout.rv_empty_post, binding.homePostRecycler.parent as ViewGroup
                )
            }
                .onSuccess { }
                .onFailure {
                    RetrofitBuilder.defaultErrorHandler(it)
                }
        }
    }

    private fun setUserInfo() {
        GlideImageLoadDelegator.loadUserImage(this, userDetailInfo.userImg, binding.homeUserimg)

        binding.homeToolbar.title = userDetailInfo.nickname
        binding.homeNickname.text = userDetailInfo.nickname
        binding.homeProfiletext.text = userDetailInfo.profileText
        binding.homeRecipecount.text = userDetailInfo.recipeCount.toString()
        binding.homeFollowercount.text = userDetailInfo.followerCount.toString()
        binding.homeFollowingcount.text = userDetailInfo.followingCount.toString()
        binding.homeFollowBtn.visibility =
            if (currentMode == UIMODE.OTHERS) View.VISIBLE else View.GONE

        binding.homeUserimgRevise.setOnClickListener { reviseProfileImage() }
        binding.homeNicknameRevise.setOnClickListener { clickReviseNickname() }
        binding.homeProfiletextRevise.setOnClickListener { clickReviseContents() }

        binding.homeRecipewrapper.setOnClickListener {
            val intent = Intent(this@HomeActivity, RecipeMyListActivity::class.java)
                .putExtra("userID", userDetailInfo.userID)
            startActivity(intent)
        }
        binding.homeFollowerwrapper.setOnClickListener {
            val intent = Intent(this@HomeActivity, FollowListActivity::class.java)
                .putExtra("MODE", FollowListActivity.UIMODE.FOLLOWER)
                .putExtra("userID", userDetailInfo.userID)
            startActivity(intent)
        }
        binding.homeFollowingwrapper.setOnClickListener {
            val intent = Intent(this@HomeActivity, FollowListActivity::class.java)
                .putExtra("MODE", FollowListActivity.UIMODE.FOLLOWING)
                .putExtra("userID", userDetailInfo.userID)
            startActivity(intent)
        }
    }

    private fun reviseProfileImage() {
        val permissions = arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)

        if (Utils.checkPermission(this, permissions)) {
            CropImage.activity()
                .setGuidelines(CropImageView.Guidelines.ON)
                .setAspectRatio(1, 1)
                .setOutputUri(imageUri)
                .setCropShape(CropImageView.CropShape.OVAL)
                .setRequestedSize(IMG_WIDTH, IMG_HEIGHT)
                .setMaxCropResultSize(IMG_WIDTH, IMG_HEIGHT)
                .start(this)
        } else ActivityCompat.requestPermissions(this, permissions, CODE_PERMISSION)
    }

    private fun clickReviseNickname() {
        MaterialDialog(this).show {
            noAutoDismiss()
            title(text = "닉네임 변경")
            input(hint = "닉네임", prefill = binding.homeNickname.text.toString())
            positiveButton(text = "확인", click = { it ->
                when {
                    it.getInputField().text == null ->
                        App.showToast("닉네임을 입력 해 주세요.")
                    it.getInputField().text.length > 12 ->
                        App.showToast("12자 이하 입력 해 주세요.")
                    else -> {
                        CoroutineScope(Dispatchers.Main).launch {
                            runCatching {
                                val res1 =
                                    accountService.checkNickname(it.getInputField().text.toString())

                                val msg = res1.body()!!["msg"].asString
                                Utils.log(msg)
                                if (msg == "ok") {
                                    binding.homeNickname.text = it.getInputField().text
                                    it.dismiss()
                                } else App.showToast("이미 존재합니다.")
                            }
                                .onSuccess { }
                                .onFailure {
                                    RetrofitBuilder.defaultErrorHandler(it)
                                }
                        }
                    }
                }
            })
            negativeButton(text = "취소")
        }
    }

    private fun clickReviseContents() {
        MaterialDialog(this).show {
            noAutoDismiss()
            title(text = "프로필 변경")
            input(
                hint = "닉네임",
                inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_MULTI_LINE,
                prefill = binding.homeProfiletext.text.toString()
            )
            positiveButton(text = "확인", click = {
                when {
                    it.getInputField().text == null ->
                        App.showToast("프로필을 입력 해 주세요.")
                    it.getInputField().text.length > 60 ->
                        App.showToast("60자 이하 입력 해 주세요.")
                    it.getInputField().lineCount > 4 ->
                        App.showToast("4줄 이하 입력 해 주세요.")
                    else -> {
                        binding.homeProfiletext.text = it.getInputField().text
                        it.dismiss()
                    }
                }
            })
            negativeButton(text = "취소")
        }
    }

    private fun updateProfile() {
        progressON(this)
        if (imageUri != null) {
            imageUrl = String.format(
                getString(R.string.format_upload_file),
                currentUserID, System.currentTimeMillis().toString()
            )
            val ref = storageReference.child(getString(R.string.storage_path_profile) + imageUrl)
            ref.putFile(imageUri!!)
                .addOnSuccessListener { updateToServer() }
        } else {
            imageUrl = userDetailInfo.userImg
            updateToServer()
        }
    }

    private fun updateToServer() {
        CoroutineScope(Dispatchers.Main).launch {
            runCatching {
                accountService.updateUser(
                    userDetailInfo.userID, imageUrl!!,
                    binding.homeNickname.text.toString(),
                    binding.homeProfiletext.text.toString()
                )

                App.showToast("업데이트 되었습니다.")
                currentOperation = OPERATION.VIEW
                reviseMenu.isVisible = true
                okMenu.isVisible = false
                binding.homeRevisegroup.visibility = View.VISIBLE
                imageUri = null
                progressOFF()
            }
                .onSuccess { }
                .onFailure {
                    RetrofitBuilder.defaultErrorHandler(it)
                }
        }
    }
}